File: public/python/096-kth-largest-element.py
# LeetCode Problem #215: Kth Largest Element in an Array
# Difficulty: Medium
# Link: https://leetcode.com/problems/kth-largest-element-in-an-array/

from typing import List
import heapq
import random

# ─────────────────────────────────────────────
# APPROACH 1: Quickselect | O(n) average time | O(1) space
# EXPLAIN: Partition around a pivot in descending order; recurse on the side containing the kth position.
# WHEN: Best average-case O(n); preferred when you want to avoid extra space.

def solve_1(nums: List[int], k: int) -> int:
    def partition(lo: int, hi: int) -> int:
        pivot = nums[hi]
        i = lo
        for j in range(lo, hi):
            if nums[j] >= pivot:  # descending partition for kth largest
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
        nums[i], nums[hi] = nums[hi], nums[i]
        return i

    def quickselect(lo: int, hi: int) -> int:
        if lo == hi:
            return nums[lo]
        p = partition(lo, hi)
        if p == k - 1:
            return nums[p]
        elif p < k - 1:
            return quickselect(p + 1, hi)
        else:
            return quickselect(lo, p - 1)

    return quickselect(0, len(nums) - 1)

# ─────────────────────────────────────────────
# APPROACH 2: Min Heap of Size K | O(n log k) time | O(k) space
# EXPLAIN: Maintain a min-heap of size k; the heap top is always the kth largest element seen so far.
# WHEN: Best when k << n; works well for streaming data.

def solve_2(nums: List[int], k: int) -> int:
    min_heap: list = []
    for num in nums:
        heapq.heappush(min_heap, num)
        if len(min_heap) > k:
            heapq.heappop(min_heap)
    return min_heap[0]

# ─────────────────────────────────────────────
# APPROACH 3: Sort Descending | O(n log n) time | O(1) space
# EXPLAIN: Sort in descending order and directly index position k-1.
# WHEN: Simplest implementation; acceptable for small arrays.

def solve_3(nums: List[int], k: int) -> int:
    nums.sort(reverse=True)
    return nums[k - 1]

# ─────────────────────────────────────────────
# APPROACH 4: Max Heap (Extract k Times) | O(n + k log n) time | O(n) space
# EXPLAIN: Build a max-heap, then pop k times to reach the kth largest.
# WHEN: When k is close to n; straightforward max-heap extraction.

def solve_4(nums: List[int], k: int) -> int:
    max_heap = [-x for x in nums]
    heapq.heapify(max_heap)
    for _ in range(k - 1):
        heapq.heappop(max_heap)
    return -max_heap[0]

# ─────────────────────────────────────────────
# APPROACH 5: heapq.nlargest | O(n log k) time | O(k) space
# EXPLAIN: Use Python's heapq.nlargest which returns the k largest elements; take the last one.
# WHEN: Most concise standard-library solution; nlargest uses a min-heap of size k internally.

def solve_5(nums: List[int], k: int) -> int:
    return heapq.nlargest(k, nums)[-1]

# Made with Bob

File: public/java/096-kth-largest-element.java
/*
 * LeetCode Problem #215: Kth Largest Element in an Array
 * Link: https://leetcode.com/problems/kth-largest-element-in-an-array/
 * Difficulty: Medium
 */
import java.util.*;

class Solution {

    // APPROACH 1: Quickselect | O(n) average time | O(log n) space
    // EXPLAIN: Partition in descending order; recurse only on the side containing position k-1.
    public int findKthLargest1(int[] nums, int k) {
        return quickselect(nums, 0, nums.length - 1, k - 1);
    }

    private int quickselect(int[] nums, int lo, int hi, int target) {
        if (lo == hi) return nums[lo];
        int p = partition(nums, lo, hi);
        if (p == target) return nums[p];
        else if (p < target) return quickselect(nums, p + 1, hi, target);
        else return quickselect(nums, lo, p - 1, target);
    }

    private int partition(int[] nums, int lo, int hi) {
        int pivot = nums[hi], i = lo;
        for (int j = lo; j < hi; j++) {
            if (nums[j] >= pivot) {
                int tmp = nums[i]; nums[i] = nums[j]; nums[j] = tmp; i++;
            }
        }
        int tmp = nums[i]; nums[i] = nums[hi]; nums[hi] = tmp;
        return i;
    }

    // APPROACH 2: Min Heap of Size K | O(n log k) time | O(k) space
    // EXPLAIN: Maintain a min-heap of size k; the top is always the kth largest.
    public int findKthLargest2(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) minHeap.poll();
        }
        return minHeap.peek();
    }

    // APPROACH 3: Sort Descending | O(n log n) time | O(1) space
    // EXPLAIN: Sort in descending order and return nums[k-1].
    public int findKthLargest3(int[] nums, int k) {
        Arrays.sort(nums);
        return nums[nums.length - k];
    }

    // APPROACH 4: Max Heap (Extract k Times) | O(n + k log n) time | O(n) space
    // EXPLAIN: Build a max-heap, then pop k times to reach the kth largest.
    public int findKthLargest4(int[] nums, int k) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        for (int num : nums) maxHeap.offer(num);
        for (int i = 0; i < k - 1; i++) maxHeap.poll();
        return maxHeap.peek();
    }

    // APPROACH 5: nth_element via partial sort | O(n) average time | O(1) space
    // EXPLAIN: Use partial nth-element selection: rearrange so element at position k-1 is correct.
    public int findKthLargest5(int[] nums, int k) {
        // Simulate nth_element with a quickselect that stops early
        int lo = 0, hi = nums.length - 1, target = k - 1;
        while (lo < hi) {
            int p = partitionDesc(nums, lo, hi);
            if (p == target) return nums[p];
            else if (p < target) lo = p + 1;
            else hi = p - 1;
        }
        return nums[lo];
    }

    private int partitionDesc(int[] nums, int lo, int hi) {
        int pivot = nums[hi], i = lo;
        for (int j = lo; j < hi; j++) {
            if (nums[j] >= pivot) {
                int tmp = nums[i]; nums[i] = nums[j]; nums[j] = tmp; i++;
            }
        }
        int tmp = nums[i]; nums[i] = nums[hi]; nums[hi] = tmp;
        return i;
    }
}

// Made with Bob
